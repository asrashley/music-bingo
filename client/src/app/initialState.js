import { initialState as cards } from '../cards/cardsSlice';
import { initialState as games } from '../games/gamesSlice';
import { initialState as tickets } from '../tickets/ticketsSlice';
import { initialState as user } from '../user/userSlice';

export const initialState = Object.freeze({
    cards,
    games,
    tickets,
    user,
});
