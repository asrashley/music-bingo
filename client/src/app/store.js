import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import { connectRouter } from 'connected-react-router';
import { createBrowserHistory } from 'history';
import { reducer as formReducer } from 'redux-form';

import cardsReducer from '../cards/cardsSlice';
import gamesReducer from '../games/gamesSlice';
import ticketsReducer from '../tickets/ticketsSlice';
import userReducer from '../user/userSlice';
import adminReducer from '../admin/adminSlice';
import usersMiddleware from '../user/userMiddleware';

export const history = createBrowserHistory();

export const store = configureStore({
  middleware: [...getDefaultMiddleware(), usersMiddleware, logger],
  reducer: {
    admin: adminReducer,
    cards: cardsReducer,
    form: formReducer,
    games: gamesReducer,
    tickets: ticketsReducer,
    router: connectRouter(history),
    user: userReducer,
  },
});
