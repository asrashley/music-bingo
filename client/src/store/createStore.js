import { configureStore } from '@reduxjs/toolkit';
import { createLogger } from 'redux-logger';
import { createRouterMiddleware, createRouterReducer } from '@lagunovsky/redux-react-router';
import log from 'loglevel';

import adminReducer from '../admin/adminSlice';
import directoriesReducer from '../directories/directoriesSlice';
import gamesReducer from '../games/gamesSlice';
import messagesReducer from '../messages/messagesSlice';
import routesReducer from '../routes/routesSlice';
import settingsReducer from '../settings/settingsSlice';
import systemReducer from '../system/systemSlice';
import ticketsReducer from '../tickets/ticketsSlice';
import userReducer from '../user/userSlice';
import usersMiddleware from '../user/userMiddleware';
import messagesMiddleware from '../messages/messagesMiddleware';

import { history } from './history';
const routerMiddleware = createRouterMiddleware(history);

export function createStore(preloadedState, logging) {
  const middleware = (getDefaultMiddleware) => {
    const mw = getDefaultMiddleware()
      .prepend(routerMiddleware)
      .concat(usersMiddleware)
      .concat(messagesMiddleware);
    if (logging === undefined) {
      logging = import.meta.env.DEV;
    }
    if (!logging) {
      return mw;
    }
    const logger = {
      debug: log.debug,
      info: log.info,
      warn: log.warn,
      error: log.error,
      log: log.debug,
    };
    const loggerMiddleware = createLogger({
      level: 'debug',
      logger
    });
    return mw.concat(loggerMiddleware);
  };
  return configureStore({
    middleware,
    reducer: {
      admin: adminReducer,
      directories: directoriesReducer,
      games: gamesReducer,
      messages: messagesReducer,
      routes: routesReducer,
      settings: settingsReducer,
      system: systemReducer,
      tickets: ticketsReducer,
      router: createRouterReducer(history),
      user: userReducer,
    },
    preloadedState
  });
}
