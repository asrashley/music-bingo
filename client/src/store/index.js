export { createStore } from './createStore';
export { initialState } from './initialState';
export { history } from './history';
