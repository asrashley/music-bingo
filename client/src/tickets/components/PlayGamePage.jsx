import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { BingoTicket } from './BingoTicket';

/* actions */
import { fetchUserIfNeeded } from '../../user/userSlice';
import { fetchTicketsIfNeeded, fetchTicketDetailIfNeeded } from '../ticketsSlice';
import { fetchGamesIfNeeded } from '../../games/gamesSlice';
import { setChecked } from '../../tickets/ticketsSlice';

/* selectors */
import { getMyGameTickets } from '../ticketsSelectors';
import { getGame } from '../../games/gamesSelectors';
import { getUser } from '../../user/userSelectors';

function deepCompareObjects(a, b) {
  if (typeof (a) !== typeof (b)) {
    return false;
  }
  if (!a && b) {
    return false;
  }
  if (a && !b) {
    return false;
  }
  for (let k in a) {
    if (a[k] !== b[k]) {
      return false;
    }
  }
  return true;
}

function deepCompareArrays(a, b) {
  if (a.length !== b.length) {
    return false;
  }
  for (let i = 0; i < a.length; ++i) {
    if (typeof (a[i]) !== typeof (b[i])) {
      return false;
    }
    if (typeof (a[i]) === 'object' && !deepCompareObjects(a[i], b[i])) {
      return false;
    } else if (a[i] !== b[i]) {
      return false;
    }
  }
  return true;
}

class PlayGamePage extends React.Component {
  static propTypes = {
    tickets: PropTypes.array.isRequired,
    game: PropTypes.object.isRequired,
    user: PropTypes.object.isRequired,
  };

  componentDidMount() {
    const { dispatch, game, tickets } = this.props;
    dispatch(fetchUserIfNeeded());
    if (game.pk > 0) {
      dispatch(fetchTicketsIfNeeded(game.pk));
      tickets.forEach(ticket => dispatch(fetchTicketDetailIfNeeded(game.pk, ticket.pk)));
    }
  }

  componentDidUpdate(prevProps, prevState) {
    const { dispatch, user, game, tickets } = this.props;
    if (user.pk !== prevProps.user.pk) {
      dispatch(fetchGamesIfNeeded());
    }
    if (game.pk > 0 && game.pk !== prevProps.game.pk) {
      dispatch(fetchTicketsIfNeeded(game.pk));
    }
    if (game.pk > 0 && !deepCompareArrays(tickets, prevProps.tickets)) {
      tickets.forEach(ticket => dispatch(fetchTicketDetailIfNeeded(game.pk, ticket.pk)));
    }
  }

  setChecked = (values) => {
    const { dispatch } = this.props;
    dispatch(setChecked(values));
  };

  render() {
    const { dispatch, game, tickets } = this.props;
    return (
      <div className="card-list">
        {tickets.length === 0 && <h2 className="warning">You need to choose a ticket to be able to play!</h2>}
        {tickets.map((ticket, idx) => <BingoTicket key={idx} ticket={ticket} game={game}
          setChecked={this.setChecked} dispatch={dispatch} download />)}
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    user: getUser(state, props),
    game: getGame(state, props),
    tickets: getMyGameTickets(state, props),
  };
};

PlayGamePage = connect(mapStateToProps)(PlayGamePage);

export {
  PlayGamePage
};
