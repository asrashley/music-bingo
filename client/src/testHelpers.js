import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { ConnectedRouter } from 'connected-react-router';
import log from 'loglevel';
import { screen, fireEvent } from '@testing-library/react';
import waitForExpect from 'wait-for-expect';

import { DisplayDialog } from './components/DisplayDialog';
import { createStore } from './store/createStore';
import { history } from './store/history';

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    store = createStore(preloadedState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return (<Provider store={store}>
      <ConnectedRouter history={history}>
        <DisplayDialog>
          {children}
        </DisplayDialog>
      </ConnectedRouter>
    </Provider>);
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

const protectedRoutes = {
  '/api/directory': true,
  '/api/game': true,
  '/api/settings': false,
  '/api/user': true,
};

export function jsonResponse(payload, status = 200) {
  const body = JSON.stringify(payload);
  return {
    body,
    status,
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': body.length
    }
  };
};

export function installFetchMocks(fetchMock, {
  loggedIn = false,
  currentAccessToken = 1,
  refreshToken = "refresh.token"
} = {}) {
  const responseModifiers = {};
  let serverStatus = null;
  const apiRequest = async (url, opts) => {
    log.trace(`apiRequest ${url}`);
    if (serverStatus !== null) {
      return serverStatus;
    }
    if (protectedRoutes[url]) {
      const { headers } = opts;
      log.trace(`Authorization = headers?.Authorization`);
      const bearer = `Bearer ${getAccessToken()}`;
      if (headers?.Authorization !== bearer) {
        log.debug(`Invalid access token "${headers?.Authorization}" !== "${bearer}"`);
        return 401;
      }
    }
    const filename = url.replace(/^\/api/, './fixtures');
    log.trace(`load ${filename}`);
    const data = await import(`${filename}.json`);
    log.trace(`${filename} = ${Object.keys(data['default']).join(',')}`);
    if (url === '/api/settings' && !loggedIn) {
      data['default'] = {
        privacy: data['default'].privacy
      };
    }
    const modFn = responseModifiers[url];
    if (modFn !== undefined) {
      return jsonResponse(modFn(url, data['default']));
    }
    return jsonResponse(data['default']);
  };
  const getAccessToken = () => `access.token.${currentAccessToken}`;
  const refreshAccessToken = (url, opts) => {
    currentAccessToken++;
    log.trace(`refreshAccessToken ${currentAccessToken}`);
    return {
      'accessToken': getAccessToken()
    };
  };
  const checkUser = async (url, opts) => {
    log.debug(`checkUser loggedIn=${loggedIn}`);
    if (serverStatus !== null) {
      log.debug(`checkUser serverStatus=${serverStatus}`);
      return serverStatus;
    }
    if (!loggedIn) {
      return 401;
    }
    const data = await import('./fixtures/user.json');
    data['default'].refreshToken = refreshToken;
    data['default'].accessToken = getAccessToken();
    return jsonResponse(data['default']);
  };
  const userDatabase = [{
    username: 'user',
    email: 'a.user@example.tld',
    password: 'mysecret'
  }];
  const loginUser = async (url, opts) => {
    if (serverStatus !== null) {
      log.debug(`loginUser status=${serverStatus}`);
      return serverStatus;
    }
    const { username, password } = JSON.parse(opts.body);
    const userIsValid = userDatabase.some(item => (
      item.username === username && item.password === password));
    log.debug(`loginUser: username="${username}" password="${password}" valid=${userIsValid}`);
    if (!userIsValid) {
      return 401;
    }
    const data = await import('./fixtures/user.json');
    data['default'].refreshToken = refreshToken;
    data['default'].accessToken = getAccessToken();
    loggedIn = true;
    return jsonResponse(data['default']);
  };
  const logoutUser = (url, opts) => {
    if (serverStatus !== null) {
      log.trace(`logoutUser status=${serverStatus}`);
      return serverStatus;
    }
    log.trace('logoutUser');
    loggedIn = false;
    return jsonResponse('Logged out');
  };
  const checkIfUserExists = (url, opts) => {
    if (serverStatus !== null) {
      log.debug(`checkIfUserExists status=${serverStatus}`);
      return serverStatus;
    }
    const { username, email } = JSON.parse(opts.body);
    const response = {
      "username": false,
      "email": false
    };
    userDatabase.forEach((item) => {
      if (item.username === username) {
        response.username = true;
      }
      if (item.email === email) {
        response.email = true;
      }
    });
    log.debug(`checkIfUserExists username=${username} email=${email} response=${JSON.stringify(response)}`);
    return jsonResponse(response);
  };

  fetchMock.config.fallbackToNetwork = false;
  fetchMock.config.warnOnFallback = true;
  log.debug(`installFetchMocks ${loggedIn}`);
  fetchMock
    .get('/api/directory', apiRequest)
    .get(/api\/directory\/\d+/, apiRequest)
    .get('/api/games', apiRequest)
    .get(/api\/game\/\d+/, apiRequest)
    .post('/api/refresh', refreshAccessToken)
    .get('/api/settings', apiRequest)
    .get('/api/user', checkUser)
    .post('/api/user', loginUser)
    .delete('/api/user', logoutUser)
    .post('/api/user/check', checkIfUserExists)
    .get('/api/users', apiRequest)
    .get('/api/user/guest', apiRequest);

  return {
    getAccessToken,
    getRefreshToken: () => refreshToken,
    isLoggedIn: () => loggedIn,
    addUser: (user) => userDatabase.push(user),
    setResponseModifier: (url, fn) => {
      responseModifiers[url] = fn;
    },
    setServerStatus: (code) => {
      serverStatus = code;
    },
    jsonResponse
  };
}

export async function setFormFields(fields) {
  fields.forEach(({ label, value, exact }) => {
    if (exact === undefined) {
      exact = true;
    }
    fireEvent.input(screen.getByLabelText(label, { exact }), {
      target: {
        value
      }
    });
  });
  await waitForExpect(() => {
    const last = fields[fields.length - 1];
    const { value } = document.querySelector(`input[name="${last.label}"`);
    expect(value).toBe(fields[fields.length - 1].value);
  });
}

