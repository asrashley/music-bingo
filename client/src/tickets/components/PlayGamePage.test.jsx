import React from 'react';
import { screen } from '@testing-library/react';
import log from 'loglevel';

import { fetchMock, renderWithProviders, installFetchMocks } from '../../testHelpers';
import { PlayGamePage } from './PlayGamePage';

import { createStore } from '../../store/createStore';
import { initialState } from '../../store/initialState';

describe('PlayGamePage component', () => {
  let apiMock = null;

  beforeEach(() => {
    apiMock = installFetchMocks(fetchMock, { loggedIn: true });
  });

  afterEach(() => {
    fetchMock.mockReset();
    log.resetLevel();
    apiMock = null;
  });

  it('shows an error message if user has not selected any tickets', async () => {
    const store = createStore({
      ...initialState,
      routes: {
        params: {
          gameId: "18-04-22-2",
          ticketPk: 3483
        }
      },
    });
    renderWithProviders(<PlayGamePage />, { store });
    await screen.findByText('You need to choose a ticket to be able to play!');
  });

  it('renders the selected game', async () => {
    const ticketPk = 3483;
    const store = createStore({
      ...initialState,
      routes: {
        params: {
          gameId: "18-04-22-2",
          ticketPk,
        },
      },
    });
    apiMock.setResponseModifier('/api/game/159/tickets', (url, tickets) => {
      return tickets.map(ticket => {
        if (ticket.pk === ticketPk) {
          return {
            ...ticket,
            user: 1
          };
        }
        return ticket;
      });
    });
    const { tracks } = await import(`../../fixtures/game/159/ticket/${ticketPk}.json`);
    //log.setLevel('debug');
    const { asFragment } = renderWithProviders(<PlayGamePage />, { store });
    await screen.findByText(tracks[0].title);
    expect(asFragment()).toMatchSnapshot();
  });
});
