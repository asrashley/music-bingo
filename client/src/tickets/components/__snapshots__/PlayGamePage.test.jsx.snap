// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PlayGamePage component > renders the selected game 1`] = `
<DocumentFragment>
  <div
    class="card-list"
  >
    <div
      class="view-ticket"
    >
      <div
        class="download"
      >
        <button
          class="btn btn-primary btn-lg"
        >
          Download ticket 1
        </button>
      </div>
      <table
        class="bingo-ticket table"
        data-game-id="18-04-22-2"
        data-ticket-number="1"
      >
        <thead>
          <tr>
            <th
              class="logo"
              colspan="5"
            />
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(218, 237, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  The Time Warp
                </p>
                <p
                  class="artist"
                >
                  Jason Donovan & The Rocky Horror Picture Show Cast
                </p>
              </div>
            </td>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(240, 248, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  Can You Feel The Love Tonight?
                </p>
                <p
                  class="artist"
                >
                  John Barrowman
                </p>
              </div>
            </td>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(218, 237, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  The Phantom Of The Opera
                </p>
                <p
                  class="artist"
                >
                  Michael Crawford Feat. Barbara Bonney
                </p>
              </div>
            </td>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(240, 248, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  Waiting For A Girl Like You
                </p>
                <p
                  class="artist"
                >
                  Constantine Maroulis
                </p>
              </div>
            </td>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(218, 237, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  My Favourite Things
                </p>
                <p
                  class="artist"
                >
                  Connie Fisher
                </p>
              </div>
            </td>
          </tr>
          <tr>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(240, 248, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  I Know Him So Well 
                </p>
                <p
                  class="artist"
                >
                  Barbara Dickson
                </p>
              </div>
            </td>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(218, 237, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  America
                </p>
                <p
                  class="artist"
                >
                  Jennifer Hepburn, Nick Ferranti & Company
                </p>
              </div>
            </td>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(240, 248, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  Talk To The Animals
                </p>
                <p
                  class="artist"
                >
                  Phillip Schofield
                </p>
              </div>
            </td>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(218, 237, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  Don't Cry For Me Argentina
                </p>
                <p
                  class="artist"
                >
                  Julie Covington
                </p>
              </div>
            </td>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(240, 248, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  I'm A Believer
                </p>
                <p
                  class="artist"
                >
                  Shrek Ensemble
                </p>
              </div>
            </td>
          </tr>
          <tr>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(218, 237, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  Pop Muzik
                </p>
                <p
                  class="artist"
                >
                  M
                </p>
              </div>
            </td>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(240, 248, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  All I Ask Of You
                </p>
                <p
                  class="artist"
                >
                  Cliff Richard & Sarah Brightman
                </p>
              </div>
            </td>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(218, 237, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  Tell Me It's Not True
                </p>
                <p
                  class="artist"
                >
                  Barbara Dickson
                </p>
              </div>
            </td>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(240, 248, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  Send In The Clowns
                </p>
                <p
                  class="artist"
                >
                  Sian Phillips With The NSO Ensemble
                </p>
              </div>
            </td>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(218, 237, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  Over The Rainbow
                </p>
                <p
                  class="artist"
                >
                  Helen Hobson
                </p>
              </div>
            </td>
          </tr>
        </tbody>
        <tfoot>
          <tr>
            <th
              class="footer-title"
              colspan="3"
            >
              Various Artists
            </th>
            <th
              class="footer-number"
              colspan="2"
            >
              Game 18-04-22-2 / Ticket 1
            </th>
          </tr>
        </tfoot>
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PlayGamePage component renders the selected game 1`] = `
<DocumentFragment>
  <div
    class="card-list"
  >
    <div
      class="view-ticket"
    >
      <div
        class="download"
      >
        <button
          class="btn btn-primary btn-lg"
        >
          Download ticket 1
        </button>
      </div>
      <table
        class="bingo-ticket table"
        data-game-id="18-04-22-2"
        data-ticket-number="1"
      >
        <thead>
          <tr>
            <th
              class="logo"
              colspan="5"
            />
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(218, 237, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  The Time Warp
                </p>
                <p
                  class="artist"
                >
                  Jason Donovan & The Rocky Horror Picture Show Cast
                </p>
              </div>
            </td>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(240, 248, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  Can You Feel The Love Tonight?
                </p>
                <p
                  class="artist"
                >
                  John Barrowman
                </p>
              </div>
            </td>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(218, 237, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  The Phantom Of The Opera
                </p>
                <p
                  class="artist"
                >
                  Michael Crawford Feat. Barbara Bonney
                </p>
              </div>
            </td>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(240, 248, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  Waiting For A Girl Like You
                </p>
                <p
                  class="artist"
                >
                  Constantine Maroulis
                </p>
              </div>
            </td>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(218, 237, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  My Favourite Things
                </p>
                <p
                  class="artist"
                >
                  Connie Fisher
                </p>
              </div>
            </td>
          </tr>
          <tr>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(240, 248, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  I Know Him So Well 
                </p>
                <p
                  class="artist"
                >
                  Barbara Dickson
                </p>
              </div>
            </td>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(218, 237, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  America
                </p>
                <p
                  class="artist"
                >
                  Jennifer Hepburn, Nick Ferranti & Company
                </p>
              </div>
            </td>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(240, 248, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  Talk To The Animals
                </p>
                <p
                  class="artist"
                >
                  Phillip Schofield
                </p>
              </div>
            </td>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(218, 237, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  Don't Cry For Me Argentina
                </p>
                <p
                  class="artist"
                >
                  Julie Covington
                </p>
              </div>
            </td>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(240, 248, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  I'm A Believer
                </p>
                <p
                  class="artist"
                >
                  Shrek Ensemble
                </p>
              </div>
            </td>
          </tr>
          <tr>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(218, 237, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  Pop Muzik
                </p>
                <p
                  class="artist"
                >
                  M
                </p>
              </div>
            </td>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(240, 248, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  All I Ask Of You
                </p>
                <p
                  class="artist"
                >
                  Cliff Richard & Sarah Brightman
                </p>
              </div>
            </td>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(218, 237, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  Tell Me It's Not True
                </p>
                <p
                  class="artist"
                >
                  Barbara Dickson
                </p>
              </div>
            </td>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(240, 248, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  Send In The Clowns
                </p>
                <p
                  class="artist"
                >
                  Sian Phillips With The NSO Ensemble
                </p>
              </div>
            </td>
            <td
              class="bingo-cell  with-artist"
              style="width: 20%; background-color: rgb(218, 237, 255);"
            >
              <div
                class="bingo-cell-wrap"
              >
                <p
                  class="title"
                >
                  Over The Rainbow
                </p>
                <p
                  class="artist"
                >
                  Helen Hobson
                </p>
              </div>
            </td>
          </tr>
        </tbody>
        <tfoot>
          <tr>
            <th
              class="footer-title"
              colspan="3"
            >
              Various Artists
            </th>
            <th
              class="footer-number"
              colspan="2"
            >
              Game 18-04-22-2 / Ticket 1
            </th>
          </tr>
        </tfoot>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
